### ISSUE TITLE
[WK06-Day026][21.09.07.Tue] RNN, LSTM, GRU

### ISSUE THREAD

**Issue by KimDaeUng** [Boostcamp-AI-Tech-2th-level01-04/TWL#21]
# Intro
- 멘토님과 짧은 만남, 논문 읽기 주제 관련 30분 톡
- 활발한 피어세션 : 꽉찬 2시간 동안 대충 넘어가는 것 없이 정리
- 필수과제 2, 3번 풀이

# 학습 내용
## RNN
- Character-level LM의 작동 예시와 이것이 할 수 있는 것들(희극 생성, c code 생성, 논문 생성) 등을 살펴봄
- Backpropagation through time(BPTT) : GPU 메모리 등 처리 용량의 한계로 전체 시퀀스를 한 번에 학습하는 대신 제한된 길이의 시퀀스 만으로 학습을 진행
- RNN의 특성 분석 연구
  - [Visualizing and Understanding Recurrent Neural Networks(Andrej et al., 2015)](https://arxiv.org/pdf/1506.02078.pdf)
  - hidden state vector의 셀 일부를 고정을 하고, 그 값이 time step이 진행됨에 따라 어떻게 변화하는지를 분석한 연구
- RNN의 Vanishing/Exploding Gradient 문제(또는 Long-term dependency 문제),
  - 원인 : 반복적으로 곱해지는 $W_{hh}$

## LSTM and GRU
- 기존 RNN을 대체
- Gradient Vanishing/Exploding 과 Long-term dependency에 대해 개선된 성능
- Long Short-Term Memory
  - hidden state vector : Short-Term Memory, 현재 time step에서 노출할 필요가 있는 정보만을 남긴 필터링된 벡터
  - cell state vector : Long-Term Memory, 장기적으로 유지하는 완전한 정보를 포함하는 벡터
- GRU
  - LSTM의 경량화 버전
  - LSTM의 cell state와 hidden state를 통합함
  - 동작원리는 동일함
- Uninterrupted gradient flow(ResNet과 유사한 구조)
  - cell state의 업데이트 과정이 $W_hh$와 같은 행렬 곱이 아닌 합과 요소별 곱 연산으로 되어 있어 backpropagation 시 gradient를 큰 병형 없이 전달

## Summary on RNN/LSTM/GRU
- 다양한 길이의 seq 길이 데이터 처리에 특화된 유연한 구조
- LSTM과 GRU를 훨씬 더 많이 사용하고, cell state vector, hidden state vector를 연산하는 방식이 기본적으로 덧셈에 기반한 방식이기 때문에 gradient vanishing/exploding 문제를 피하고 long-term dependency 문제를 해결할 수 있다.

# 과제 수행
- 필수과제 2
